pkg/indexer/indexer.go

```diff
@@ -0,0 +1,418 @@
+package indexer
+
+import (
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+	"gorm.io/driver/sqlite"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+
+	"github.com/gohornet/hornet/pkg/model/milestone"
+	"github.com/gohornet/hornet/pkg/model/utxo"
+	"github.com/gohornet/hornet/pkg/whiteflag"
+	"github.com/iotaledger/hive.go/kvstore/utils"
+	iotago "github.com/iotaledger/iota.go/v3"
+)
+
+var (
+	ErrNotFound = errors.New("output not found for given filter")
+
+	tables = []interface{}{
+		&status{},
+		&extendedOutput{},
+		&nft{},
+		&foundry{},
+		&alias{},
+	}
+)
+
+type Indexer struct {
+	db *gorm.DB
+}
+
+func NewIndexer(dbPath string) (*Indexer, error) {
+
+	if err := utils.CreateDirectory(dbPath, 0700); err != nil {
+		return nil, err
+	}
+
+	dbFile := filepath.Join(dbPath, "indexer.db")
+
+	db, err := gorm.Open(sqlite.Open(dbFile), &gorm.Config{})
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the tables and indexes if needed
+	if err := db.AutoMigrate(tables...); err != nil {
+		return nil, err
+	}
+
+	return &Indexer{
+		db: db,
+	}, nil
+}
+
+func processSpent(spent *utxo.Spent, tx *gorm.DB) error {
+	switch spent.OutputType() {
+	case iotago.OutputExtended:
+		return tx.Where("output_id = ?", spent.OutputID()[:]).Delete(&extendedOutput{}).Error
+	case iotago.OutputAlias:
+		return tx.Where("output_id = ?", spent.OutputID()[:]).Delete(&alias{}).Error
+	case iotago.OutputNFT:
+		return tx.Where("output_id = ?", spent.OutputID()[:]).Delete(&nft{}).Error
+	case iotago.OutputFoundry:
+		return tx.Where("output_id = ?", spent.OutputID()[:]).Delete(&foundry{}).Error
+	}
+	return nil
+}
+
+func processOutput(output *utxo.Output, tx *gorm.DB) error {
+	switch iotaOutput := output.Output().(type) {
+	case *iotago.ExtendedOutput:
+		features, err := iotaOutput.FeatureBlocks().Set()
+		if err != nil {
+			return err
+		}
+
+		conditions, err := iotaOutput.UnlockConditions().Set()
+		if err != nil {
+			return err
+		}
+
+		extended := &extendedOutput{
+			OutputID:       make(outputIDBytes, iotago.OutputIDLength),
+			Amount:         iotaOutput.Amount,
+			MilestoneIndex: output.MilestoneIndex(),
+		}
+		copy(extended.OutputID, output.OutputID()[:])
+
+		if senderBlock := features.SenderFeatureBlock(); senderBlock != nil {
+			extended.Sender, err = addressBytesForAddress(senderBlock.Address)
+			if err != nil {
+				return err
+			}
+		}
+
+		if tagBlock := features.TagFeatureBlock(); tagBlock != nil {
+			copy(extended.Tag, tagBlock.Tag)
```

Comment:
Does this copy work if the `extended.Tag` is not initialized with the correct size?

Reply:
I'll add `nft.Tag = make([]byte, len(tagBlock.Tag))` before the copy